generic-ec-curves/src/utils.rs

```diff
@@ -0,0 +1,270 @@
+        }
+        32 => {
+            #[allow(clippy::expect_used)]
+            let bytes: &[u8; 32] = bytes.try_into().expect("we checked that bytes len == 32");
+            S::from_le_array_mod_order(bytes)
+        }
+        33.. => {
+            let two_to_256 = S::add(&S::from_le_array_mod_order(&[0xff; 32]), one);
+
+            let chunks = bytes.chunks_exact(32);
+            let remainder = if !chunks.remainder().is_empty() {
+                Some(scalar_from_le_bytes_mod_order_reducing_32::<S>(
+                    chunks.remainder(),
+                    one,
+                ))
+            } else {
+                None
+            };
+
+            let chunks = chunks.rev().map(|chunk| {
+                #[allow(clippy::expect_used)]
+                let chunk: &[u8; 32] = chunk.try_into().expect("wrong chunk size");
+                S::from_le_array_mod_order(chunk)
+            });
+
+            remainder
+                .into_iter()
+                .chain(chunks)
+                .reduce(|acc, int| S::add(&S::mul(&acc, &two_to_256), &int))
+                .unwrap_or_default()
+        }
+    }
+}
+
+/// Interprets `bytes` as big-endian encoding of an integer, takes it modulo curve (prime)
+/// order and returns scalar `S`
+///
+/// Works with scalars for which only [`Reduce<32>`][Reduce] is defined.
+///
+/// Takes:
+/// * Big-endian `bytes` representation of the integer
+/// * Scalar `one = 1`
+pub fn scalar_from_be_bytes_mod_order_reducing_32<S>(bytes: &[u8], one: &S) -> S
+where
+    S: Default + Copy,
+    S: Reduce<32>,
+    S: generic_ec_core::Additive + generic_ec_core::Multiplicative<S, Output = S>,
+{
+    let len = bytes.len();
+    match len {
+        ..=31 => {
+            let mut padded = [0u8; 32];
+            padded[32 - len..].copy_from_slice(bytes);
+            S::from_be_array_mod_order(&padded)
+        }
+        32 => {
+            #[allow(clippy::expect_used)]
+            let bytes: &[u8; 32] = bytes.try_into().expect("we checked that bytes len == 32");
+            S::from_be_array_mod_order(bytes)
+        }
+        33.. => {
+            let two_to_256 = S::add(&S::from_be_array_mod_order(&[0xff; 32]), one);
+
+            let chunks = bytes.rchunks_exact(32);
+            let remainder = if !chunks.remainder().is_empty() {
+                Some(scalar_from_be_bytes_mod_order_reducing_32::<S>(
+                    chunks.remainder(),
+                    one,
+                ))
+            } else {
+                None
+            };
+
+            let chunks = chunks.rev().map(|chunk| {
+                #[allow(clippy::expect_used)]
+                let chunk: &[u8; 32] = chunk.try_into().expect("wrong chunk size");
+                S::from_be_array_mod_order(chunk)
+            });
+
+            remainder
+                .into_iter()
+                .chain(chunks)
+                .reduce(|acc, int| S::add(&S::mul(&acc, &two_to_256), &int))
+                .unwrap_or_default()
```

Comment:
What do you think to refactor this part into a separate function? It differs in 32/64, chunks/rchunks, from_be_array_mod_order/from_le_array_mod_order, and which function to recurse to.

Reply:
You could even replace the recursing call with from_le_array_mod_order

Reply:
Here we indeed can remove the recursion call, I will do that. But you can't remove recursion in `32_64` call, which shows that we can't move it into separate function and reuse across 32 and 32_64

Reply:
Challenge accepted

Reply:
```rust
fn reduce_chunks<'a, const N: usize, F, I, S>(
    is_be: bool,
    remainder: &'a [u8],
    chunks: I,
    from_array: F,
    one: &S,
) -> S
where
    F: Fn(&[u8; N]) -> S,
    I: DoubleEndedIterator<Item = &'a [u8]>,
    S: Default,
    S: generic_ec_core::Additive + generic_ec_core::Multiplicative<S, Output = S>,
{
    let modulo = S::add(&from_array(&[0xff; N]), one);
    let remainder = if !remainder.is_empty() {
        let mut padded = [0u8; N];
        let len = remainder.len();
        if is_be {
            padded[N - len..].copy_from_slice(remainder);
        } else {
            padded[..len].copy_from_slice(remainder);
        }
        Some(from_array(&padded))
    } else {
        None
    };
    let chunks = chunks.rev().map(|chunk| {
        #[allow(clippy::expect_used)]
        let chunk: &[u8; N] = chunk.try_into().expect("wrong chunk size");
        from_array(chunk)
    });

    remainder
        .into_iter()
        .chain(chunks)
        .reduce(|acc, int| S::add(&S::mul(&acc, &modulo), &int))
        .unwrap_or_default()
}
```
used like
```rust
        33.. => {
            let chunks = bytes.rchunks_exact(32);
            reduce_chunks(
                true,
                chunks.remainder(),
                chunks,
                |a: &[u8; 32]| S::from_be_array_mod_order(a),
                one,
            )
        }
```

Reply:
That would only work for `32` tho, it won't work for `32_64`

Reply:
In `32_64` remainder can fit into either 32 or 64 bytes array. If it fits into 32 bytes array, conversion is cheaper, so we should use that.

Reply:
I just found out that lifetimes in `impl Fn` are now supported, so you can eve make the declaration 2 (or 3) lines shorter

Reply:
Imo it's getting too complex, I rather leave it as is

Reply:
I disagree, but I do think we're not going to touch this code anymore, unless we decide to rewrite it completely, so ok